### ISSUE TITLE
[DLCV]2018-02-03 질문사항

### ISSUE THREAD

**Issue by Jooong** [YBIGTA/data-science-2018#10]
[Face Recognition](https://jooong.github.io/ml/2018/02/02/Face-recognition/) and [Neural Style Transfer](https://jooong.github.io/ml/2018/02/02/Neural-style-transefer/)

**Comment by statkim7578** [comment]
![1231222](https://user-images.githubusercontent.com/34841210/35740883-8ca45d38-0879-11e8-9bf2-7d11ccf3946a.PNG)

임의의 이미지를 conv-layer에 통과시키면 activation map이 나온다고 하셨는데요!
activation map을 conv-layer에 다시 집어넣기도 하는건가요..? 무슨일이 생기는거죠 그러면?

**Comment by godongyoung** [comment]
1) 생성 이미지 G까지 gradient를 역전파시키는 부분에서 '네트워크는 얼려둔 상태'라는건  pre-trained ConvNet을 의미하는 건가요? 그때 style cost를 계산하는 convnet과 content cost를 계산하는 convnet은 같은 (pre-trained) model인건가요?

2)![Imgur](https://i.imgur.com/lTzmUGu.png)
이게 진짭니다ㅠ Gram matrix간의 mean-squared distance가 왜 저렇게 나뉘어 질까요?

**Comment by projokim** [comment]
https://camo.githubusercontent.com/5fbea1e33a64c5442bf175a29ec5644093a246db/68747470733a2f2f692e696d6775722e636f6d2f777478486b774a2e706e67
9개의 이미지를 9개의 필터로 convolution한 것인가요??

**Comment by jungsooyun** [comment]
![default](https://user-images.githubusercontent.com/32008845/35748999-623e83a6-0893-11e8-9297-078b7443d2e0.JPG)
Triplet을 만드는 과정이 궁금해서 병규 블로그를 찾아봤는데 Mini-batch를 이용해서 그 안에서 데이터를 찾았는데 엄청나게 큰(?) 배치사이즈를 사용했다고 서술되어 있더라고요. 이 부분이 이해가 안되는데 설명해주실 수 있나요!!

**Comment by hyejukim-ybigta** [comment]
margin(alpha)는 보통 얼마로 설정하나요?

**Comment by Jooong** [comment]
@statkim7578 일반적으로 CNN 모델을 쌓을 때 conv-layer를 여러층 쌓게 됩니다. 이전 레이어의 activation을 conv-layer에 넣어준다는 의미였습니다.
![image](https://user-images.githubusercontent.com/19547969/35759886-d1327746-08bf-11e8-9874-3a1cd5604aa4.png)
@godongyoung 1. 그게 그거입니다. (강의에 따르면) style transfer는 이를 위한 특별한 네트워크를 설계하지 않습니다. 이미 학습되어있는 CNN 모델을 이용합니다. 2. 우선 summation 안에 제곱이 빠져있어서 수정했습니다. [여기](http://forums.fast.ai/t/quick-math-question-for-the-style-loss-function-in-artistic-neural-style-transfer/2881)에 따르면 미분값이 이쁘게 나오라고 붙여준 것 같고, 실제 학습에 있어서 영향은 없다고 이야기합니다.
@projokim 엄밀히 말하자면 9개의 필터를 먼저 고르고, 이후 해당 필터에서 가장 highly-activated images를 9개 골라 시각화한 것입니다.
@jungsooyun 전체 데이터셋 안에서 이미지별로 hard-negative를 찾는 것은 굉장히 time-consuming한 작업입니다. 때문에 미니배치 안에서 hard-negative를 찾는 방법을 택했습니다. 대신, distance의 신뢰도(?!) 안정성(?!)을 보장하고자 미니배치 사이즈를 크게 둔 것이죠.
@hyejukim-ybigta 0.2정도 둡니다.

**Comment by DabinJeong** [comment]
Neural Style transfer에서 cost fuction을 계산할 때 왜 중간 레이어까지만 forward하는 건가요??

**Comment by Jooong** [comment]
@Rafaela94 기본적으로 vgg는 classifier이기 때문에 뒷단에 fc나 softmax가 달려있습니다. 이 부분은 이미지의 feature를 학습하기보다는, 학습된 feature를 활용하여 이미지를 분류하는 역할을 합니다. 따라서 이미지 feature 정보를 학습하는 중간단의 conv-layer를 활용하여 content 혹은 style cost를 계산합니다. 뒷부분은 필요가 없기 때문에 중간 레이어까지만 forward합니다.

**Comment by ybigtaJH** [comment]
Siamese network 에 대해서 좀 더 설명해주셨으면 좋겠습니다

**Comment by SongChiYoung** [comment]
> 먼저, 생성 이미지 G를 랜덤하게 initialize한다. 이후 계산된 cost인 J(G)" role="presentation" style="box-sizing: border-box; -webkit-tap-highlight-color: transparent; display: inline; font-style: normal; font-weight: normal; line-height: normal; font-size: 16px; text-indent: 0px; text-align: left; text-transform: none; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; direction: ltr; max-width: none; max-height: none; min-width: 0px; min-height: 0px; border: 0px; padding: 0px; margin: 0px; color: rgb(51, 51, 51); font-family: Lato, Sans-serif; font-variant-ligatures: normal; font-variant-caps: normal; orphans: 2; widows: 2; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); position: relative;">J(G)J(G)를 backpropagate한다. 생성 이미지 G까지 gradient를 역전파시키고, 이를 이용하여 G를 update한다. 네트워크는 얼려둔 상태에서 G를 직접 update하는 형태다.

부분이 잘 이해가 되지 않습니다. ㅜ

**Comment by godongyoung** [comment]
![Imgur](https://i.imgur.com/lTzmUGu.png)
저기서 l은 content를 계산할때의 레이어와 같나요, 아님 전체 레이어로 계산하나요?
또 보통 vgg19에서 content레이어를 계산하는 l은 몇번째로 하나요??

**Comment by Jooong** [comment]
@godongyoung 선택하기 나름인 것 같습니다. 원 논문에서는 Style을 1: (conv1_1), 2: (conv1_1,conv2_1), 3: (conv1_1, conv2_1, conv3_1), 4: (conv1_1,...,conv4_1), 5: (conv1_1, ... , conv5_1) 다섯가지 set 중에서 하나를 골라서 썼구요, content 같은 경우에는 conv1_1, conv2_1, conv3_1, conv4_1, conv5_1 중 하나를 골라서 써보았다고 합니다. 강의 예제에서는 content든 style이든 conv4_2를 사용했습니다.
